### PR TITLE
fix: allow trace id to switch within the same connection

### PR DESCRIPTION
**Pull request by GaoleMeng**
This change will allow the trace id to be changeable within the same connection. So that the same connection pool could be used for multiple different trace id from different stream writer. We will pass a new trace id to server side once we detect a table switch

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/java-bigquerystorage/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> ☕️

If you write sample code, please follow the [samples format](
https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md).

### CODE REVIEW

File: google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/ConnectionWorker.java [hunk 1]
```diff
@@ -1372,16 +1375,21 @@ static final class AppendRequestAndResponse {
     // If a response is no longer expected this is set back to null.
     Instant requestSendTimeStamp;
 
+    // Mark where the traffic coming from. Used for internal logging purpose.
+    String fullTraceId;
```
Comment (agrawal-siddharth) [inline google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/ConnectionWorker.java:1379]: Why not retrieve the traceId from streamWriter.getFullTraceId()?

File: google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/ConnectionWorkerTest.java [hunk 2]
```diff
@@ -97,11 +97,13 @@ public void testMultiplexedAppendSuccess() throws Exception {
           StreamWriter.newBuilder(TEST_STREAM_1, client)
               .setWriterSchema(createProtoSchema("foo"))
               .setLocation("us")
+              .setTraceId("some_head:trace_id_1")
               .build();
       StreamWriter sw2 =
           StreamWriter.newBuilder(TEST_STREAM_2, client)
               .setWriterSchema(createProtoSchema("complicate"))
               .setLocation("us")
+              .setTraceId("some_head:trace_id_2")
```
Comment (agrawal-siddharth) [inline google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/ConnectionWorkerTest.java:106]: Do you also have a test where the StreamWriter traceId is left as NULL?

**Review by agrawal-siddharth — COMMENTED** [review]
Although this is OK for now, eventually we will also want to update the traceId on the telemetryMetrics when we have a table switch. Since currently we only use traceId for Dataflow in the telemetryMetrics, and Dataflow does not attempt to change the traceId across streams, we don't need this now.

File: google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/ConnectionWorker.java (hunk 1 above)
Comment (GaoleMeng) [reply google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/ConnectionWorker.java:1379]: ConnectionWorker does not have reference to individual stream writer

Comment (GaoleMeng) [reply google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/ConnectionWorker.java:1379]: Since multiple SW can share the same connecctionWorker

File: google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/ConnectionWorkerTest.java (hunk 2 above)
Comment (GaoleMeng) [reply google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/ConnectionWorkerTest.java:106]: Done

File: google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/ConnectionWorker.java (hunk 1 above)
Comment (agrawal-siddharth) [reply google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/ConnectionWorker.java:1379]: I meant the streamWriter found in AppendRequestAndResponse. This is available for each request, right?

Comment (GaoleMeng) [reply google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/ConnectionWorker.java:1379]: You are right, done

**Review by agrawal-siddharth — APPROVED** [review]
